File: scripts/parse-menu-from-zip.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import zipfile
import os
import json
import shutil
from pathlib import Path
import re

# Пути
ZIP_FILE = Path(__file__).parent.parent / "sapiens photo.zip"
EXTRACT_DIR = Path(__file__).parent.parent / "temp_extracted"
MENU_DIR = Path(__file__).parent.parent / "src" / "assets" / "menu"
MENU_JSON_PATH = Path(__file__).parent.parent / "menu.json"

# Категории для автоматической классификации
CATEGORY_KEYWORDS = {
    'Десерты': ['десерт', 'пирог', 'торт', 'кекс', 'вафля', 'блинчик', 'сырник', 'чизкейк', 'медовик', 'синнабон', 'крафл', 'орео', 'варенье'],
    'Закуски': ['закуск', 'оливк', 'маслин', 'артишок', 'карпаччо', 'брускетт'],
    'Мясные блюда': ['мясн', 'перепелк', 'утк', 'котлет', 'шатобриан', 'брискет', 'бургер', 'бекон', 'окорок', 'омлет', 'яйц'],
    'Рыба и морепродукты': ['рыб', 'лосос', 'тунец', 'угор', 'креветк', 'гребешок', 'краб', 'икра', 'ролл', 'суши', 'голубец', 'треск', 'щук', 'темпура', 'нори'],
    'Салаты': ['салат', 'руккола', 'боул', 'коул', 'stefan'],
    'Супы': ['суп', 'бульон', 'том-ям', 'вонтон'],
    'Суши и роллы': ['ролл', 'суши', 'калифорни', 'филадельфи', 'радуга'],
    'Завтраки': ['завтрак', 'вафля', 'бриошь', 'драник', 'скрэмбл', 'птитим', 'киноа', 'овсян', 'сырник'],
    'Прочее': []
}

def detect_category(dish_name):
    """Определяет категорию блюда по названию"""
    lower_name = dish_name.lower()
    
    for category, keywords in CATEGORY_KEYWORDS.items():
        for keyword in keywords:
            if keyword in lower_name:
                return category
    
    return 'Прочее'

def normalize_filename(file_name):
    """Нормализует имя файла (удаляет расширение)"""
    # Удаляем расширение
    name = re.sub(r'\.(jpg|jpeg|png|JPG|JPEG|PNG)$', '', file_name, flags=re.IGNORECASE)
    return name.strip()

def get_raw_filename_bytes(zip_path, member_name):
    """Получает сырые байты имени файла из ZIP архива"""
    import struct
    with open(zip_path, 'rb') as f:
        data = f.read()
    
    pos = 0
    while True:
        pos = data.find(b'PK\x01\x02', pos)
        if pos == -1:
            break
        
        try:
            name_len = struct.unpack('<H', data[pos + 28:pos + 30])[0]
            filename_start = pos + 46
            filename_bytes = data[filename_start:filename_start + name_len]
            
            # Проверяем, соответствует ли это нашему файлу
            # Сравниваем с оригинальным именем
            test_decoded = filename_bytes.decode('cp437', errors='ignore')
            if member_name in test_decoded or test_decoded[:30] in member_name[:50]:
                return filename_bytes
        except:
            pass
        
        pos += 1
        if pos >= len(data) - 100:
            break
    
    return None

def fix_filename_encoding(filename, zip_path=None):
    """Исправляет кодировку имени файла из ZIP архива"""
    # Если передан путь к ZIP, пытаемся получить сырые байты
    if zip_path:
        raw_bytes = get_raw_filename_bytes(zip_path, filename)
        if raw_bytes:
            # Пробуем декодировать через cp866 (правильная кодировка для русских имен в ZIP)
            try:
                decoded = raw_bytes.decode('cp866', errors='ignore')
                return decoded
            except:
                pass
    
    # Fallback: пробуем исправить уже декодированное имя
    try:
        if isinstance(filename, str):
            # Пытаемся представить как байты и передекодировать
            # Если имя уже неправильно декодировано, это сложнее исправить
            pass
    except Exception as e:
        pass
    
    return filename

def extract_zip_with_encoding(zip_path, extract_dir):
    """Извлекает zip-архив с правильной обработкой кодировки имен файлов"""
    import struct
    
    if extract_dir.exists():
        shutil.rmtree(extract_dir)
    extract_dir.mkdir(parents=True, exist_ok=True)
    
    images_info = []
    
    # Сначала читаем ZIP файл как бинарный, чтобы получить правильные имена
    filename_map = {}  # маппинг неправильного имени -> правильного
    
    with open(zip_path, 'rb') as f:
        data = f.read()
    
    pos = 0
    while True:
        pos = data.find(b'PK\x01\x02', pos)
        if pos == -1:
            break
        
        try:
            name_len = struct.unpack('<H', data[pos + 28:pos + 30])[0]
            filename_start = pos + 46
            filename_bytes = data[filename_start:filename_start + name_len]
            
            # Декодируем через cp866 (правильная кодировка)
            correct_name = filename_bytes.decode('cp866', errors='ignore')
            # Также получаем имя, как его видит zipfile (через cp437)
            incorrect_name = filename_bytes.decode('cp437', errors='ignore')
            
            if correct_name != incorrect_name:
                filename_map[incorrect_name] = correct_name
        except:
            pass
        
        pos += 1
        if pos >= len(data) - 100:
            break
    
    # Теперь извлекаем файлы
    with zipfile.ZipFile(zip_path, 'r') as zip_ref:
        for member in zip_ref.namelist():
            if member.endswith('/'):
                continue
            
            # Получаем правильное имя
            fixed_name = filename_map.get(member, member)
            
            # Извлекаем файл
            try:
                # Извлекаем с оригинальным именем (member)
                zip_ref.extract(member, extract_dir)
                
                # Получаем путь без директории "sapiens photo/"
                if 'sapiens photo/' in fixed_name:
                    relative_path = fixed_name.replace('sapiens photo/', '')
                else:
                    relative_path = os.path.basename(fixed_name)
                
                old_path = extract_dir / member
                new_path = extract_dir / relative_path
                
                # Создаем директорию для нового пути, если нужно
                new_path.parent.mkdir(parents=True, exist_ok=True)
                
                if old_path.exists():
                    # Удаляем файл, если он уже существует
                    if new_path.exists() and old_path != new_path:
                        new_path.unlink()
                    # Переименовываем файл
                    if old_path != new_path:
                        old_path.rename(new_path)
                        print(f"✓ {relative_path[:60]}...")
                    
                    # Сохраняем информацию об изображении
                    if any(relative_path.lower().endswith(ext) for ext in ['.jpg', '.jpeg', '.png']):
                        images_info.append((new_path, relative_path))
            except Exception as e:
                print(f"✗ Ошибка при извлечении {member}: {e}")
    
    return images_info

def find_images(directory):
    """Рекурсивно находит все изображения в директории"""
    image_extensions = ['.jpg', '.jpeg', '.png']
    images = []
    
    for root, dirs, files in os.walk(directory):
        for file in files:
            if any(file.lower().endswith(ext) for ext in image_extensions):
                images.append(Path(root) / file)
    
    return images

def create_safe_filename(name, extension):
    """Создает безопасное имя файла"""
    # Заменяем небезопасные символы на подчеркивания
    safe_name = re.sub(r'[^\w\s-]', '_', name)
    safe_name = re.sub(r'[-\s]+', '_', safe_name)
    return f"{safe_name}{extension}"

def main():
    print('Начинаю парсинг меню из zip-архива...')
    
    # Извлекаем zip-архив
    print('Извлекаю файлы из zip-архива...')
    extracted_images = extract_zip_with_encoding(ZIP_FILE, EXTRACT_DIR)
    
    # Если функция вернула список изображений, используем его
    if extracted_images:
        images = extracted_images
        print(f'Найдено {len(images)} изображений')
    else:
        # Иначе ищем все изображения рекурсивно
        print('Ищу изображения...')
        images = find_images(EXTRACT_DIR)
        print(f'Найдено {len(images)} изображений')
    
    # Читаем существующий menu.json
    existing_menu = {
        'menu': {'categories': []},
        'all_items': [],
        'statistics': {'total_items': 0, 'categories_count': 0}
    }
    max_id = 0
    
    if MENU_JSON_PATH.exists():
        try:
            with open(MENU_JSON_PATH, 'r', encoding='utf-8') as f:
                existing_menu = json.load(f)
            # Находим максимальный ID
            if existing_menu.get('all_items'):
                max_id = max([item.get('id', 0) for item in existing_menu['all_items']], default=0)
        except Exception as e:
            print(f'Не удалось прочитать существующий menu.json: {e}, создаю новый')
    
    # Создаем карту существующих блюд
    existing_dishes = {}
    if existing_menu.get('all_items'):
        for item in existing_menu['all_items']:
            existing_dishes[item.get('name', '').lower()] = item
    
    # Убеждаемся, что директория menu существует
    MENU_DIR.mkdir(parents=True, exist_ok=True)
    
    # Обрабатываем каждое изображение
    new_dishes = []
    category_map = {}
    
    for item in images:
        # images теперь содержит кортежи (путь, правильное имя)
        if isinstance(item, tuple):
            image_path, correct_filename = item
            file_name = correct_filename
        else:
            image_path = item
            file_name = image_path.name
        
        dish_name = normalize_filename(file_name)
        
        # Пропускаем, если имя пустое или слишком короткое
        if not dish_name or len(dish_name) < 3:
            print(f'Пропускаю файл с неподходящим именем: {file_name}')
            continue
        
        # Проверяем, не существует ли уже такое блюдо
        if dish_name.lower() in existing_dishes:
            print(f'Блюдо "{dish_name}" уже существует, пропускаю')
            continue
        
        # Определяем расширение
        ext = image_path.suffix.lower()
        image_format = ext[1:]  # убираем точку
        
        # Создаем безопасное имя файла
        safe_file_name = create_safe_filename(dish_name, ext)
        target_image_path = MENU_DIR / safe_file_name
        
        # Копируем изображение
        shutil.copy2(image_path, target_image_path)
        print(f'Скопировано: {file_name} -> {safe_file_name}')
        
        # Определяем категорию
        category = detect_category(dish_name)
        
        # Создаем объект блюда
        max_id += 1
        dish = {
            'id': max_id,
            'name': dish_name,
            'category': category,
            'image': f'images/{safe_file_name}',
            'image_format': image_format,
            'description': None,
            'composition': None,
            'allergens': None
        }
        
        new_dishes.append(dish)
        
        # Добавляем в карту категорий
        if category not in category_map:
            category_map[category] = []
        category_map[category].append(dish)
    
    print(f'\nОбработано {len(new_dishes)} новых блюд')
    print(f'Категории: {", ".join(category_map.keys())}')
    
    # Объединяем существующие и новые блюда
    all_dishes = (existing_menu.get('all_items') or []) + new_dishes
    
    # Группируем по категориям
    categories_map = {}
    
    # Сначала добавляем существующие категории
    if existing_menu.get('menu') and existing_menu['menu'].get('categories'):
        for cat in existing_menu['menu']['categories']:
            categories_map[cat['name']] = cat.get('items', [])
    
    # Добавляем новые блюда к категориям
    for dish in new_dishes:
        if dish['category'] not in categories_map:
            categories_map[dish['category']] = []
        categories_map[dish['category']].append(dish)
    
    # Формируем структуру меню
    menu_structure = {
        'menu': {
            'categories': [
                {
                    'name': category_name,
                    'items': items,
                    'count': len(items)
                }
                for category_name, items in categories_map.items()
            ]
        },
        'all_items': all_dishes,
        'statistics': {
            'total_items': len(all_dishes),
            'categories_count': len(categories_map)
        }
    }
    
    # Сохраняем menu.json
    with open(MENU_JSON_PATH, 'w', encoding='utf-8') as f:
        json.dump(menu_structure, f, ensure_ascii=False, indent=2)
    
    print(f'\nmenu.json обновлен: {len(all_dishes)} блюд в {len(categories_map)} категориях')
    
    # Очищаем временную директорию
    if EXTRACT_DIR.exists():
        shutil.rmtree(EXTRACT_DIR)
    
    print('\nГотово!')

if __name__ == '__main__':
    main()

File: scripts/parse-menu-from-zip.js
const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const stream = require('stream');
const pipeline = promisify(stream.pipeline);

// Используем yauzl для работы с zip (если нужно будет установить)
// Пока используем простой подход с child_process для unzip
const { execSync } = require('child_process');

const ZIP_FILE = path.join(__dirname, '..', 'sapiens photo.zip');
const EXTRACT_DIR = path.join(__dirname, '..', 'temp_extracted');
const MENU_DIR = path.join(__dirname, '..', 'src', 'assets', 'menu');
const MENU_JSON_PATH = path.join(__dirname, '..', 'menu.json');

// Категории для автоматической классификации
const CATEGORY_KEYWORDS = {
  'Десерты': ['десерт', 'пирог', 'торт', 'кекс', 'вафля', 'блинчик', 'сырник', 'чизкейк', 'медовик', 'синнабон', 'крафл', 'орео'],
  'Закуски': ['закуск', 'оливк', 'маслин', 'артишок', 'карпаччо', 'брускетт'],
  'Мясные блюда': ['мясн', 'перепелк', 'утк', 'котлет', 'шатобриан', 'брискет', 'бургер', 'бекон', 'окорок', 'омлет', 'яйц'],
  'Рыба и морепродукты': ['рыб', 'лосос', 'тунец', 'угор', 'креветк', 'гребешок', 'краб', 'икра', 'ролл', 'суши', 'голубец', 'треск', 'щук'],
  'Салаты': ['салат', 'руккола', 'боул', 'коул'],
  'Супы': ['суп', 'бульон', 'том-ям', 'вонтон'],
  'Суши и роллы': ['ролл', 'суши', 'калифорни', 'филадельфи'],
  'Завтраки': ['завтрак', 'вафля', 'бриошь', 'драник', 'скрэмбл', 'птитим', 'киноа', 'овсян'],
  'Прочее': []
};

// Функция для определения категории по названию блюда
function detectCategory(dishName) {
  const lowerName = dishName.toLowerCase();
  
  for (const [category, keywords] of Object.entries(CATEGORY_KEYWORDS)) {
    for (const keyword of keywords) {
      if (lowerName.includes(keyword)) {
        return category;
      }
    }
  }
  
  return 'Прочее';
}

// Функция для нормализации имени файла (удаление расширения и исправление кодировки)
function normalizeFileName(fileName) {
  // Удаляем расширение
  let name = fileName.replace(/\.(jpg|jpeg|png|JPG|JPEG|PNG)$/i, '');
  
  // Пытаемся исправить кодировку, если она неправильная
  // Если имя уже в правильной кодировке, просто возвращаем
  try {
    // Пытаемся декодировать как CP1251 и перекодировать в UTF-8
    if (Buffer.from(name, 'latin1').toString('utf8') !== name) {
      name = Buffer.from(name, 'latin1').toString('utf8');
    }
  } catch (e) {
    // Если не получается, оставляем как есть
  }
  
  return name.trim();
}

// Основная функция парсинга
async function parseMenuFromZip() {
  console.log('Начинаю парсинг меню из zip-архива...');
  
  // Очищаем временную директорию
  if (fs.existsSync(EXTRACT_DIR)) {
    fs.rmSync(EXTRACT_DIR, { recursive: true, force: true });
  }
  fs.mkdirSync(EXTRACT_DIR, { recursive: true });
  
  // Извлекаем zip-архив
  console.log('Извлекаю файлы из zip-архива...');
  try {
    execSync(`unzip -q -o "${ZIP_FILE}" -d "${EXTRACT_DIR}"`, { encoding: 'utf8' });
  } catch (error) {
    console.error('Ошибка при извлечении zip:', error.message);
    // Попробуем использовать node-yauzl или другой метод
    console.log('Пробую альтернативный метод...');
    // Используем python для извлечения с правильной кодировкой
    const pythonScript = `
import zipfile
import sys
import os

zip_path = sys.argv[1]
extract_dir = sys.argv[2]

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    for member in zip_ref.namelist():
        try:
            # Пытаемся исправить кодировку имени файла
            try:
                name = member.encode('cp437').decode('cp1251')
            except:
                name = member
            zip_ref.extract(member, extract_dir)
            # Переименовываем файл с правильной кодировкой
            old_path = os.path.join(extract_dir, member)
            new_path = os.path.join(extract_dir, name)
            if os.path.exists(old_path) and old_path != new_path:
                os.makedirs(os.path.dirname(new_path), exist_ok=True)
                if os.path.exists(new_path):
                    os.remove(new_path)
                os.rename(old_path, new_path)
        except Exception as e:
            print(f"Ошибка при обработке {member}: {e}", file=sys.stderr)
`;
    fs.writeFileSync(path.join(EXTRACT_DIR, 'extract.py'), pythonScript);
    try {
      execSync(`python3 "${path.join(EXTRACT_DIR, 'extract.py')}" "${ZIP_FILE}" "${EXTRACT_DIR}"`, { encoding: 'utf8' });
    } catch (pyError) {
      console.error('Ошибка при использовании Python:', pyError.message);
      throw new Error('Не удалось извлечь zip-архив');
    }
  }
  
  // Ищем все изображения во всех поддиректориях
  console.log('Ищу изображения...');
  const imageExtensions = ['.jpg', '.jpeg', '.png', '.JPG', '.JPEG', '.PNG'];
  const images = [];
  
  function findImages(dir) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      
      if (entry.isDirectory()) {
        findImages(fullPath);
      } else if (imageExtensions.some(ext => entry.name.endsWith(ext))) {
        images.push(fullPath);
      }
    }
  }
  
  findImages(EXTRACT_DIR);
  
  console.log(`Найдено ${images.length} изображений`);
  
  // Читаем существующий menu.json, если он есть
  let existingMenu = { menu: { categories: [] }, all_items: [], statistics: { total_items: 0, categories_count: 0 } };
  let maxId = 0;
  
  if (fs.existsSync(MENU_JSON_PATH)) {
    try {
      existingMenu = JSON.parse(fs.readFileSync(MENU_JSON_PATH, 'utf8'));
      // Находим максимальный ID
      if (existingMenu.all_items && existingMenu.all_items.length > 0) {
        maxId = Math.max(...existingMenu.all_items.map(item => item.id || 0));
      }
    } catch (e) {
      console.warn('Не удалось прочитать существующий menu.json, создаю новый');
    }
  }
  
  // Создаем карту существующих блюд по имени для проверки дубликатов
  const existingDishes = new Map();
  if (existingMenu.all_items) {
    existingMenu.all_items.forEach(item => {
      existingDishes.set(item.name.toLowerCase(), item);
    });
  }
  
  // Обрабатываем каждое изображение
  const newDishes = [];
  const categoryMap = new Map();
  
  // Убеждаемся, что директория menu существует
  if (!fs.existsSync(MENU_DIR)) {
    fs.mkdirSync(MENU_DIR, { recursive: true });
  }
  
  for (const imagePath of images) {
    const fileName = path.basename(imagePath);
    const dishName = normalizeFileName(fileName);
    
    // Пропускаем, если имя пустое или слишком короткое
    if (!dishName || dishName.length < 3) {
      console.warn(`Пропускаю файл с неподходящим именем: ${fileName}`);
      continue;
    }
    
    // Проверяем, не существует ли уже такое блюдо
    if (existingDishes.has(dishName.toLowerCase())) {
      console.log(`Блюдо "${dishName}" уже существует, пропускаю`);
      continue;
    }
    
    const ext = path.extname(imagePath).toLowerCase();
    const imageFormat = ext.replace('.', '');
    
    // Создаем безопасное имя файла для сохранения
    const safeFileName = `${dishName.replace(/[^a-zA-Z0-9а-яёА-ЯЁ\s]/g, '_').replace(/\s+/g, '_')}${ext}`;
    const targetImagePath = path.join(MENU_DIR, safeFileName);
    
    // Копируем изображение
    fs.copyFileSync(imagePath, targetImagePath);
    console.log(`Скопировано: ${fileName} -> ${safeFileName}`);
    
    // Определяем категорию
    const category = detectCategory(dishName);
    
    // Создаем объект блюда
    maxId++;
    const dish = {
      id: maxId,
      name: dishName,
      category: category,
      image: `images/${safeFileName}`,
      image_format: imageFormat,
      description: null,
      composition: null,
      allergens: null
    };
    
    newDishes.push(dish);
    
    // Добавляем в карту категорий
    if (!categoryMap.has(category)) {
      categoryMap.set(category, []);
    }
    categoryMap.get(category).push(dish);
  }
  
  console.log(`\nОбработано ${newDishes.length} новых блюд`);
  console.log(`Категории: ${Array.from(categoryMap.keys()).join(', ')}`);
  
  // Объединяем существующие и новые блюда
  const allDishes = [...(existingMenu.all_items || []), ...newDishes];
  
  // Группируем по категориям
  const categoriesMap = new Map();
  
  // Сначала добавляем существующие категории
  if (existingMenu.menu && existingMenu.menu.categories) {
    existingMenu.menu.categories.forEach(cat => {
      categoriesMap.set(cat.name, [...cat.items]);
    });
  }
  
  // Добавляем новые блюда к категориям
  newDishes.forEach(dish => {
    if (!categoriesMap.has(dish.category)) {
      categoriesMap.set(dish.category, []);
    }
    categoriesMap.get(dish.category).push(dish);
  });
  
  // Формируем структуру меню
  const menuStructure = {
    menu: {
      categories: Array.from(categoriesMap.entries()).map(([categoryName, items]) => ({
        name: categoryName,
        items: items,
        count: items.length
      }))
    },
    all_items: allDishes,
    statistics: {
      total_items: allDishes.length,
      categories_count: categoriesMap.size
    }
  };
  
  // Сохраняем menu.json
  fs.writeFileSync(MENU_JSON_PATH, JSON.stringify(menuStructure, null, 2), 'utf8');
  console.log(`\nmenu.json обновлен: ${allDishes.length} блюд в ${categoriesMap.size} категориях`);
  
  // Очищаем временную директорию
  fs.rmSync(EXTRACT_DIR, { recursive: true, force: true });
  
  console.log('\nГотово!');
}

// Запускаем парсинг
parseMenuFromZip().catch(error => {
  console.error('Ошибка:', error);
  process.exit(1);
});

